### PR TITLE
drivers/at86rf2xx: code size optimisations

### PR DESCRIPTION
**Pull request by thomaseichinger**
With #4412 I was concerned that a PR adding features to one kind of at86rf2xx supported devices, increases code size for all others by 100 bytes.
This PR tries to reduce the increase to it's bare minimum i.e. get/set of options.

~~Based on #4412.~~

### CODE REVIEW

**Comment by daniel-k** [comment]
@thomaseichinger I won't be able to review this until in 2 weeks, so maybe assign someone else if you want to get this merged for release 2015.12. Sorry :(

**Comment by thomaseichinger** [comment]
Added another commit to use `assert` for netdev's get/set functions to check for type sizes.

File: drivers/include/at86rf2xx.h [hunk 1]
```diff
@@ -89,14 +90,15 @@ extern "C" {
...
  */
-#define AT86RF2XX_STATE_TRX_OFF      (0x08)     /**< idle */
-#define AT86RF2XX_STATE_PLL_ON       (0x09)     /**< ready to transmit */
-#define AT86RF2XX_STATE_SLEEP        (0x0f)     /**< sleep mode */
-#define AT86RF2XX_STATE_BUSY_RX_AACK (0x11)     /**< busy receiving data */
-#define AT86RF2XX_STATE_BUSY_TX_ARET (0x12)     /**< busy transmitting data */
-#define AT86RF2XX_STATE_RX_AACK_ON   (0x16)     /**< wait for incoming data */
-#define AT86RF2XX_STATE_TX_ARET_ON   (0x19)     /**< ready for sending data */
-#define AT86RF2XX_STATE_IN_PROGRESS  (0x1f)     /**< ongoing state conversion */
+#define AT86RF2XX_STATE_FORCE_TRX_OFF  (0x03)     /**< force transition to idle */
```
Comment (DipSwitch) [inline drivers/include/at86rf2xx.h]: This is only a command and not a state, so `_set_state` should be changed to something like this:

    #define _set_state(dev, state)        _set_state_ex(dev, state, state)

	static inline void _set_state_ex(at86rf2xx_t *dev, uint8_t cmd, uint8_t state)
	{
		at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_STATE, cmd);
		while (at86rf2xx_get_status(dev) != state);
		dev->state = state;
	}

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 2]
```diff
@@ -446,6 +427,11 @@ void at86rf2xx_set_state(at86rf2xx_t *dev, uint8_t state)
     if (state == old_state) {
         return;
     }
+
+    if (state == AT86RF2XX_STATE_FORCE_TRX_OFF) {
+        _set_state(dev, state);
```
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: _set_state_ex(dev, AT86RF2XX_STATE_FORCE_TRX_OFF, AT86RF2XX_STATE_TRX_OFF)

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 3]
```diff
@@ -469,7 +455,7 @@ void at86rf2xx_set_state(at86rf2xx_t *dev, uint8_t state)
 
     if (state == AT86RF2XX_STATE_SLEEP) {
         /* First go to TRX_OFF */
-        _force_trx_off(dev);
+        at86rf2xx_set_state(dev, AT86RF2XX_STATE_FORCE_TRX_OFF);
```
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: _set_state_ex(dev, AT86RF2XX_STATE_FORCE_TRX_OFF, AT86RF2XX_STATE_TRX_OFF)

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 4]
```diff
@@ -491,5 +477,5 @@ void at86rf2xx_reset_state_machine(at86rf2xx_t *dev)
         old_state = at86rf2xx_get_status(dev);
     } while (old_state == AT86RF2XX_STATE_IN_PROGRESS);
 
-    _force_trx_off(dev);
+    at86rf2xx_set_state(dev, AT86RF2XX_STATE_FORCE_TRX_OFF);
```
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: _set_state_ex(dev, AT86RF2XX_STATE_FORCE_TRX_OFF, AT86RF2XX_STATE_TRX_OFF)

**Comment by DipSwitch** [comment]
Wait ignore that commends, that's not the proper fix ^_^ since the `_set_state` is not public

File: drivers/at86rf2xx/at86rf2xx_getset.c (hunk 3 above)
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: This is correct.

File: drivers/at86rf2xx/at86rf2xx_getset.c (hunk 4 above)
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: This is correct aswell.

File: drivers/at86rf2xx/at86rf2xx_getset.c (hunk 2 above)
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: This should indeed be changed :)

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 5]
```diff
@@ -428,31 +418,28 @@ static int _get(gnrc_netdev_t *device, netopt_t opt, void *val, size_t max_len)
...
-            if (max_len < sizeof(uint16_t)) {
-                return -EOVERFLOW;
-            }
+            assert(max_len < sizeof(uint16_t));
             ((uint8_t *)val)[1] = 0;
             ((uint8_t *)val)[0] = at86rf2xx_get_chan(dev);
             return sizeof(uint16_t);
 
+        case NETOPT_CHANNEL_PAGE:
+            assert(max_len < sizeof(uint8_t));
```
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: By changing the size this should also be updated in

 `sys/shell/commands/sc_netif.c#228`

    res = gnrc_netapi_get(dev, NETOPT_CHANNEL_PAGE, 0, &u16, sizeof(u16));

    if (res >= 0) {
        printf(" Page: %" PRIu16 " ", u16);
    }

to:

    res = gnrc_netapi_get(dev, NETOPT_CHANNEL_PAGE, 0, &u8, sizeof(u8));

    if (res >= 0) {
        printf(" Page: %" PRIu8 " ", u8);
    }

 `sys/shell/commands/sc_netif.c#574`
`return _netif_set_u16(dev, NETOPT_CHANNEL_PAGE, value);` to `return _netif_set_u8(dev, NETOPT_CHANNEL_PAGE, value);`

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 6]
```diff
@@ -608,92 +583,87 @@ static int _set(gnrc_netdev_t *device, netopt_t opt, void *val, size_t len)
...
             }
+            at86rf2xx_set_chan(dev, chan);
+            res = sizeof(uint16_t);
             break;
 
-        case NETOPT_TX_POWER:
-            if (len > sizeof(int16_t)) {
-                res = -EOVERFLOW;
+        case NETOPT_CHANNEL_PAGE:
+            assert(len != sizeof(uint8_t));
```
Comment (DipSwitch) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Also affects my previous comment

**Comment by DipSwitch** [comment]
I tried to create a pull request to your brance but I couldn't for some reason. The complete patch to make it working

	diff --git a/drivers/at86rf2xx/at86rf2xx_getset.c b/drivers/at86rf2xx/at86rf2xx_getset.c
	index 130f6f6..46e2ed3 100644
	--- a/drivers/at86rf2xx/at86rf2xx_getset.c
	+++ b/drivers/at86rf2xx/at86rf2xx_getset.c
	@@ -413,9 +413,11 @@ void at86rf2xx_set_option(at86rf2xx_t *dev, uint16_t option, bool state)
		 }
	 }
	 
	-static inline void _set_state(at86rf2xx_t *dev, uint8_t state)
	+#define _set_state(dev, state)        _set_state_ex(dev, state, state)
	+
	+static inline void _set_state_ex(at86rf2xx_t *dev, uint8_t cmd, uint8_t state)
	 {
	-    at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_STATE, state);
	+    at86rf2xx_reg_write(dev, AT86RF2XX_REG__TRX_STATE, cmd);
		 while (at86rf2xx_get_status(dev) != state);
		 dev->state = state;
	 }
	@@ -429,7 +431,7 @@ void at86rf2xx_set_state(at86rf2xx_t *dev, uint8_t state)
		 }
	 
		 if (state == AT86RF2XX_STATE_FORCE_TRX_OFF) {
	-        _set_state(dev, state);
	+        _set_state_ex(dev, state, AT86RF2XX_STATE_TRX_OFF);
			 return;
		 }
		 /* make sure there is no ongoing transmission, or state transition already
	diff --git a/sys/shell/commands/sc_netif.c b/sys/shell/commands/sc_netif.c
	index 97a71a0..fe2a6aa 100644
	--- a/sys/shell/commands/sc_netif.c
	+++ b/sys/shell/commands/sc_netif.c
	@@ -225,10 +225,10 @@ static void _netif_list(kernel_pid_t dev)
			 printf(" Channel: %" PRIu16 " ", u16);
		 }
	 
	-    res = gnrc_netapi_get(dev, NETOPT_CHANNEL_PAGE, 0, &u16, sizeof(u16));
	+    res = gnrc_netapi_get(dev, NETOPT_CHANNEL_PAGE, 0, &u8, sizeof(u8));
	 
		 if (res >= 0) {
	-        printf(" Page: %" PRIu16 " ", u16);
	+        printf(" Page: %" PRIu8 " ", u8);
		 }
	 
		 res = gnrc_netapi_get(dev, NETOPT_NID, 0, &u16, sizeof(u16));
	@@ -571,7 +571,7 @@ static int _netif_set(char *cmd_name, kernel_pid_t dev, char *key, char *value)
			 return _netif_set_u16(dev, NETOPT_CHANNEL, value);
		 }
		 else if (strcmp("page", key) == 0) {
	-        return _netif_set_u16(dev, NETOPT_CHANNEL_PAGE, value);
	+        return _netif_set_u8(dev, NETOPT_CHANNEL_PAGE, value);
		 }
		 else if ((strcmp("nid", key) == 0) || (strcmp("pan", key) == 0) ||
				  (strcmp("pan_id", key) == 0)) {

File: drivers/include/at86rf2xx.h (hunk 1 above)
Comment (thomaseichinger) [inline drivers/include/at86rf2xx.h]: @DipSwitch I think you have a point here, still I am sceptic to use you proposed approach. The `FORCE_TRX_OFF` command is and shout only be used internally by the drive. How about the following, (i) remove the `AT86RF2XX_STATE_FORCE_TRX_OFF` from the `at86rf2xx.h` header and use the command defined in `at86rf2xx_registers.h` and (ii) properly comment why we use it in `at86rf2xx_reset_state_machine`.

**Comment by thomaseichinger** [comment]
@DipSwitch I've omitted the type size changes. There would have to be changes in other places to, which requires an additional PR I think.

**Comment by thomaseichinger** [comment]
rebased upon merge of #4412

File: drivers/include/at86rf2xx.h (hunk 1 above)
Comment (thomaseichinger) [inline drivers/include/at86rf2xx.h]: @DipSwitch Sorry, I confused things. Adapted `_set_state` function to use a 3rd parameter. Should be sound now.

**Comment by thomaseichinger** [comment]
rebased with #3954 being merged.

**Comment by DipSwitch** [comment]
Sorry, buzy days. But I'll check again tomorrow :)

**Comment by OlegHahm** [comment]
@DipSwitch, @haukepetersen, willing to take another look?

**Comment by authmillenon** [comment]
Though this is older and has less dependencies I would really be happy if this PR is merged after #4646.

**Comment by OlegHahm** [comment]
Needs a rebase after #4646 got merged.

**Comment by authmillenon** [comment]
@thomaseichinger ping?

**Comment by OlegHahm** [comment]
Sorry, feature freeze.

**Comment by authmillenon** [comment]
Can I start a discussion if optimization is a feature or a bug fix :P?

**Comment by OlegHahm** [comment]
Optimization is always a bugfix for embedded systems. ;)

**Comment by authmillenon** [comment]
Then why postpone it because of the feature freeze. Bugfixes can always be amended ;-)

**Comment by OlegHahm** [comment]
True.

**Comment by thomaseichinger** [comment]
Rebased.

**Comment by haukepetersen** [comment]
please squash

**Comment by kYc0o** [comment]
Even if it's a bugfix, it needs to be squashed and ACKed by someone, so if no reaction today it will be postponed anyways...

**Comment by haukepetersen** [comment]
Changes look sane, and everything I tried out still runs fine. ACK once squashed.

@thomaseichinger: please squash until tonight so we can get this into the release, otherwise I will postpone this...

**Comment by haukepetersen** [comment]
postponed (though hopefully merged soon!)

**Comment by kYc0o** [comment]
Please rebase. I hope this can be integrated soon, for now it will marked for the next release.

**Comment by miri64** [comment]
Postponed due to feature freeze (and please rebase)

**Comment by thomaseichinger** [comment]
Rebased, I'm sorry for the delay and hope everything went ok - didn't have time to test yet.

**Comment by PeterKietzmann** [comment]
I will have a look at this now

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 7]
```diff
@@ -398,95 +376,49 @@ static int _set(netdev2_t *netdev, netopt_t opt, void *val, size_t len)
...
-            break;
-
-        case NETOPT_NID:
-            if (len > sizeof(uint16_t)) {
-                res = -EOVERFLOW;
-            }
-            else {
-                at86rf2xx_set_pan(dev, *((uint16_t *)val));
-                /* don't set res to set netdev2_ieee802154_t::pan */
-            }
```
Comment (PeterKietzmann) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Where are the addresses/ the pan actually written to the device?

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 8]
```diff
@@ -398,95 +376,49 @@ static int _set(netdev2_t *netdev, netopt_t opt, void *val, size_t len)
...
+        case NETOPT_CHANNEL:
+            assert(len != sizeof(uint16_t));
+            uint8_t chan = ((uint8_t *)val)[0];
+            if (chan < AT86RF2XX_MIN_CHANNEL ||
+                chan > AT86RF2XX_MAX_CHANNEL) {
+                res = -ENOTSUP;
```
Comment (PeterKietzmann) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Why not staying with `EINVAL`?

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 9]
```diff
@@ -398,95 +376,49 @@ static int _set(netdev2_t *netdev, netopt_t opt, void *val, size_t len)
...
-        case NETOPT_NID:
-            if (len > sizeof(uint16_t)) {
-                res = -EOVERFLOW;
-            }
-            else {
-                at86rf2xx_set_pan(dev, *((uint16_t *)val));
-                /* don't set res to set netdev2_ieee802154_t::pan */
-            }
+        case NETOPT_CHANNEL:
+            assert(len != sizeof(uint16_t));
```
Comment (PeterKietzmann) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Should be `sizeof(uint8_t)` right?

File: drivers/at86rf2xx/at86rf2xx_netdev.c (hunk 7 above)
Comment (thomaseichinger) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: @PeterKietzmann Thanks, good catch! Should be fixed now.

File: drivers/at86rf2xx/at86rf2xx_netdev.c (hunk 9 above)
Comment (thomaseichinger) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Changed.

File: drivers/at86rf2xx/at86rf2xx_netdev.c (hunk 8 above)
Comment (thomaseichinger) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: Changed.

**Comment by PeterKietzmann** [comment]
@thomaseichinger ACK. Please squash. I've tested several parameters, e.g. channels, short and long addresses, nid, states (without monitoring consumption) with `gnrc_networking` and `default`. Everything worked as expected. 

Btw: with `default` and without `DEVELHELP` this PR saves ~1,8kb in text segment :-)

**Comment by OlegHahm** [comment]
Nice!

**Comment by thomaseichinger** [comment]
@PeterKietzmann squashed. Thanks a million for testing!

**Pull request merged**